experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts

```diff
@@ -0,0 +1,193 @@
+						// Setting a key's value to `undefined` is equivalent to omitting the entry.
+						const filteredKeys = unfilteredKeys.filter(
+							(key) => node.get(key) !== undefined,
+						);
+						return filteredKeys;
+					} else {
+						// Assume record-like object
+						const objectedNode = node as Record<FieldKey, unknown>;
+
+						const unfilteredKeys = Object.keys(node) as FieldKey[];
+						// Object proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the key/value pair.
+						const filteredKeys = unfilteredKeys.filter((key) => {
+							const value = objectedNode[key];
+							return value !== undefined;
+						});
+						return filteredKeys;
+					}
+				default:
+					return [];
+			}
+		},
+		getFieldFromNode: (node: TNode, key: FieldKey): readonly TNode[] => {
+			if (node === undefined) {
+				throw new Error(
+					"Cursor adapter was passed an undefined node. This is unsupported.",
+				);
+			}
+
+			// Object.prototype.hasOwnProperty can return true for strings (ex: with key "0"), so we have to filter them out.
+			// Rather than just special casing strings, we can handle them with an early return for all primitives.
+			if (typeof node !== "object") {
+				return [];
+			}
+
+			if (node === null) {
+				return [];
+			}
+
+			if (Array.isArray(node)) {
+				return key === EmptyKey || key === rootFieldKey ? node : [];
+			}
+
+			if (node instanceof Map) {
+				const value = (node as Map<FieldKey, TNode>).get(key);
+				return value === undefined ? [] : [value];
+			}
+
+			if (Object.prototype.hasOwnProperty.call(node, key)) {
+				const field = (node as Record<FieldKey, TNode>)[key];
+				return field === undefined ? [] : [field];
+			}
+
+			return [];
+		},
+	};
+}
+
+/**
+ * Construct a {@link CursorWithNode} from a {@link ProxyNode}.
+ *
+ * @param node - The tree being inserted, which will be mapped to the editable tree domain via {@link createProxyTreeAdapter}.
+ * @param context - The {@link TreeEntity.context} of the parent node under which the tree is being inserted.
+ * @param typeSet - The types allowed under the parent node under which the tree is being inserted.
+ * @param mode - The mode of cursor creation.
+ * Determines whether the cursor is initialized in "node" mode or "field" mode.
+ */
+export function cursorFromProxyTree<TNode extends ProxyNode<TreeNodeSchema, "javaScript">>(
+	node: TNode,
+	context: TreeDataContext,
+	typeSet: TreeTypeSet,
+	mode: CursorLocationType = CursorLocationType.Nodes,
+): CursorWithNode<TNode> {
+	const adapter = createProxyTreeAdapter<TNode>(context, typeSet);
+	switch (mode) {
+		case CursorLocationType.Nodes:
+			return stackTreeNodeCursor(adapter, node);
+		case CursorLocationType.Fields:
+			return stackTreeFieldCursor(adapter, node);
+		default:
+			throw new Error(`Unrecognized CursorLocationType: "${mode}."`);
```

Comment:
use unreachableCase for defaults when matches should be exhastive to get compile time checking.